src/V3Delayed.cpp

```diff
@@ -76,140 +114,174 @@
 VL_DEFINE_DEBUG_FUNCTIONS;
 
 //######################################################################
-// Delayed state, as a visitor of each AstNode
+// Convert AstAssignDlys (NBAs)
 
 class DelayedVisitor final : public VNVisitor {
     // TYPES
 
-    // Components of an AstAssignDly LHS expression, e.g.:
-    //   Sel(ArraySel(ArraySel(_: VarRef, _: Index), _: Index), _: Lsb, _: Width))
-    struct LhsComponents final {
-        AstVarRef* refp = nullptr;  // The referenced variable
-        std::vector<AstNodeExpr*> arrIdxps;  // The array indices applied to 'refp'
-        AstNodeExpr* selLsbp = nullptr;  // The bit select LSB expression, after the array selects
-        AstConst* selWidthp = nullptr;  // The width of the bit select
+    // The various NBA conversion schemes, including error cases
+    enum class Scheme : uint8_t {
+        Undecided = 0,
+        UnsupportedCompoundArrayInLoop,
+        ShadowVar,
+        FlagShared,
+        FlagUnique,
+        ValueQueueWhole,
+        ValueQueuePartial
+    };
+
+    // All info associated with a variable that is the target of an NBA
+    class VarScopeInfo final {
```

Comment:
Consider moving out of the visitor to a top-level class?  Ok if not.

Reply:
This is the "user1p" of AstVarScope, and in all other algorithms in the code base we keep the definition in the relevant visitor, so I left it.